src/probnum/kernels/_kernel.py

```diff
@@ -0,0 +1,273 @@
+            containing pairwise evaluations for all observations in ``x0`` (and ``x1``).
+        """
+        raise NotImplementedError
+
+    @property
+    def input_dim(self) -> int:
+        """Dimension of arguments of the covariance function.
+
+        The dimension of inputs to the covariance function :math:`k : \\mathbb{R}^{
+        d_{in}} \\times \\mathbb{R}^{d_{in}} \\rightarrow
+        \\mathbb{R}^{d_{out} \\times d_{out}}`.
+        """
+        return self._input_dim
+
+    @property
+    def output_dim(self) -> int:
+        """Dimension of the evaluated covariance function.
+
+        The resulting evaluated kernel :math:`k(x_0, x_1) \\in
+        \\mathbb{R}^{d_{out} \\times d_{out}}` has *shape=(output_dim,
+        output_dim)*.
+        """
+        return self._output_dim
+
+    def _check_and_reshape_inputs(
+        self,
+        x0: _InputType,
+        x1: Optional[_InputType] = None,
+    ) -> Tuple[np.ndarray, Union[np.ndarray, None], ShapeType]:
+        """Check and transform inputs of the covariance function.
+
+        Checks the shape of the inputs to the covariance function and
+        transforms the inputs into two-dimensional :class:`numpy.ndarray`s such that
+        inputs are stacked row-wise.
+
+        Parameters
+        ----------
+        x0 :
+            First input to the covariance function.
+        x1 :
+            Second input to the covariance function.
+
+        Returns
+        -------
+        x0 :
+            First input to the covariance function.
+        x1 :
+            Second input to the covariance function.
+        kernshape :
+            Shape of the evaluation of the covariance function.
+
+        Raises
+        -------
+        ValueError :
+            If input shapes of x0 and x1 do not match the kernel input dimension or
+            each other.
+        """
+        # pylint: disable="too-many-boolean-expressions"
+        # Transform into array(s) and add second argument
+        x0 = np.asarray(x0)
+        equal_inputs = False
+        if x1 is None:
+            x1 = x0
+            equal_inputs = True
+        else:
+            x1 = np.asarray(x1)
+
+        # Error message
+        err_msg = (
+            f"Argument shapes x0.shape={x0.shape} and x1.shape="
+            f"{x1.shape} do not match kernel input dimension "
+            f"{self.input_dim}. Try passing either two vectors or two matrices with the"
+            f" second dimension equalling the kernel input dimension."
+        )
+
+        # Check and promote shapes
+        if x1 is None:
+
+            if (
+                (x0.ndim == 0 and self.input_dim > 1)  # Scalar input
+                or (x0.ndim == 1 and x0.shape[0] != self.input_dim)  # Vector input
+                or (x0.ndim >= 2 and x0.shape[1] != self.input_dim)  # Matrix input
+            ):
+                raise ValueError(err_msg)
+            equal_inputs = True
+        else:
+            # Promote unequal shapes
```

Comment:
Pinging the author here :) This is no real review, rather a good practice question I may need to adopt :)

Reply:
wait, have I asked the same question twice? There is a very similar thread below. If so, please ignore :D

Reply:
I have thought about this quite a bit and I am not sure what the best practice is we should adopt. Python functions since it is dynamically typed seems to encourage functions to be vectorized and not care about input shape much. Then again modifying input arguments may be unintuitive to the user. I think we have to discuss this in the software meeting and decide on a few things related to this question. Personally I have struggled a lot with `GPy` initially when I tried the library first simply because it was always expecting 2d row vectors when evaluating the kernel function.